docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,331 @@
+
+The Associated Token Program needs to support all token programs seamlessly.
+Currently, it performs the following sequence:
+
+* `SystemInstruction::Transfer` the required lamports to make the SPL holding account rent-exempt
+* `SystemInstruction::Allocate` space for the SPL holding account
+* `SystemInstruction::Assign` the holding account to the SPL token program
+* `InitializeHolding` on SPL token program
+
+This does not work for an opaque token program, because we do not know the size required
+from the outside. Conversely, if we allow a token program to take the lamports it wants,
+a malicious token program could take too much from users.
+
+To get around this, the Token Program interface has a new `CreateHolding`
+instruction, which only allocates space and assigns the account to the program.
+Once the space is allocated, the Associated Token Program transfers the
+required lamports, and returns an error if that number is too large. The process
+becomes:
+
+* call `CreateHolding` on the token program (allocate and assign)
+* calculate rent requirement based on the data size of holding account
+* `SystemInstruction::Transfer` the required lamports to make the holding account rent-exempt
+* call `InitializeHolding` on the token program
+
+### Web3 / Wallets
+
+To properly support i-tokens, wallets and applications will need to make the
+following changes:
+
+* get holdings: query for holdings owned by the public key, for all official
+token programs listed in the registry (not just `Tokenkeg...`)
+* get balances: avoid deserializing holding data, and instead use
+`getBalance` from RPC
+* transfer tokens: only use `ApproveChecked` and `TransferChecked` on the token
+program
+* create holding: avoid directly creating an account and initializing,
+and use the associated token account program instead
+
+### On-chain Programs
+
+To properly support i-tokens on-chain, programs will need to make these changes:
+
+* multiple token programs: all programs that support SPL Tokens must always
+include the program id of the token(s) holdings they're passing in.
+For example, token-swap needs to accept separate token programs for token A and B
+* hard-coded token program: avoid using `Tokenkeg...` directly, and delete `id()`
+the SPL token code
+* get balance: instead of deserializing the account data into an SPL holding, use
+the new wrapper to deserialize or call `GetHoldingBalance` instruction with the
+appropriate token program
+* get owner: same as above, but using the `GetHoldingOwner`
+* get supply: same as above, but using the `GetMintSupply`
+* get mint authority: same as above, but using the `GetMintAuthority`
+* get mint decimals: same as above, but using the `GetMintDecimals`
+* transfer: only use `TransferChecked`, which requires passing mints
+
+## Other New Token Programs
+
+For awhile now, people have been coming up with ideas that likely require a new
+token program to implement. These include:
+
+### Voting Tokens
```

Comment:
how about adding stake-pool's tokens with regard to its rewards?

Reply:
I'm not sure if we need a new token program for that -- an oracle on the stake pool seems like a better approach to me, but I might be missing something.  How would this work?

Reply:
never mind, then. ;) I'm basically noob about the stake pool. I felt stake-rewards are like compounding-interest-link thing so, I wondered stake-pool _tokens_ could be nicely modelled around this i-token.